crates/shared/src/price_estimation/trade_verifier.rs

```diff
@@ -553,68 +733,203 @@ enum Error {
+            sell_token => BigRational::from_integer((-500).into()),
+            buy_token => BigRational::from_integer(0.into()),
+        };
+
         let sell_less = SettleOutput {
             gas_used: 0.into(),
             out_amount: 2_000.into(),
-            buy_tokens_lost: BigRational::from_integer(0.into()),
-            sell_tokens_lost: BigRational::from_integer((-500).into()),
+            tokens_lost,
         };
         // Ending up with surplus in the buffers is always fine
-        let estimate = ensure_quote_accuracy(&low_threshold, &query, H160::zero(), &sell_less);
+        let estimate = ensure_quote_accuracy(&low_threshold, &query, &trade, &sell_less);
         assert!(estimate.is_ok());
 
+        let tokens_lost = hashmap! {
+            sell_token => BigRational::from_integer(0.into()),
+            buy_token => BigRational::from_integer((-1_000).into()),
+        };
+
         let pay_out_less = SettleOutput {
             gas_used: 0.into(),
             out_amount: 2_000.into(),
-            buy_tokens_lost: BigRational::from_integer((-1_000).into()),
-            sell_tokens_lost: BigRational::from_integer(0.into()),
+            tokens_lost,
         };
         // Ending up with surplus in the buffers is always fine
-        let estimate = ensure_quote_accuracy(&low_threshold, &query, H160::zero(), &pay_out_less);
+        let estimate = ensure_quote_accuracy(&low_threshold, &query, &trade, &pay_out_less);
         assert!(estimate.is_ok());
     }
+
+    #[test]
+    fn ensure_quote_accuracy_with_jit_orders() {
+        // Inaccuracy limit of 10%
+        let low_threshold = big_decimal_to_big_rational(&BigDecimal::from_str("0.1").unwrap());
+        let high_threshold = big_decimal_to_big_rational(&BigDecimal::from_str("0.11").unwrap());
+
+        let sell_token: H160 = H160::from_low_u64_be(1);
+        let buy_token: H160 = H160::from_low_u64_be(2);
+
+        let mut clearing_prices = HashMap::new();
+        clearing_prices.insert(sell_token, U256::from(1u64));
+        clearing_prices.insert(buy_token, U256::from(2u64));
+
+        let query_with_out_amount = |kind: OrderKind| -> (PriceQuery, U256) {
+            let (in_amount, out_amount) = match kind {
+                OrderKind::Sell => (500u64.into(), 250u64.into()),
+                OrderKind::Buy => (250u64.into(), 500u64.into()),
+            };
+            let price_query = PriceQuery {
+                in_amount: NonZeroU256::new(in_amount).unwrap(),
+                kind,
+                sell_token,
+                buy_token,
+            };
+
+            (price_query, out_amount)
+        };
+
+        // The jit order should fully fill the query
+        let jit_executed_amount = |side: &dto::Side| -> U256 {
+            match side {
+                dto::Side::Sell => 250u64.into(),
+                dto::Side::Buy => 500u64.into(),
+            }
+        };
+
+        let jit_order = |side: dto::Side| -> dto::JitOrder {
```

Comment:
AFAICS `ensure_quote_accuracy()` only needs the `solver` from the trade kind and not the whole `TradeKind`. If you were to revert that change back the test setup should become a lot simpler.